transports/webrtc/src/udp_mux.rs

```diff
@@ -0,0 +1,372 @@
+                let conn = {
+                    let address_map = self.address_map.read();
+                    address_map.get(&addr).map(Clone::clone)
+                };
+
+                let conn = match conn {
+                    // If we couldn't find the connection based on source address, see if
+                    // this is a STUN mesage and if so if we can find the connection based on ufrag.
+                    None if is_stun_message(read.filled()) => {
+                        self.conn_from_stun_message(&read.filled(), &addr)
+                    }
+                    s @ Some(_) => s,
+                    _ => None,
+                };
+
+                match conn {
+                    None => {
+                        if !self.new_addrs.read().contains(&addr) {
+                            match ufrag_from_stun_message(read.filled(), false) {
+                                Ok(ufrag) => {
+                                    log::trace!(
+                                        "Notifying about new address addr={} from ufrag={}",
+                                        &addr,
+                                        ufrag
+                                    );
+                                    let mut new_addrs = self.new_addrs.write();
+                                    new_addrs.insert(addr);
+                                    return Poll::Ready(UDPMuxEvent::NewAddr(NewAddr {
+                                        addr,
+                                        ufrag,
+                                    }));
+                                }
+                                Err(e) => {
+                                    log::debug!(
+                                        "Unknown address addr={} (non STUN packet: {})",
+                                        &addr,
+                                        e
+                                    );
+                                }
+                            }
+                        }
+                    }
+                    Some(conn) => {
+                        let mut packet = Vec::with_capacity(read.filled().len());
+                        packet.copy_from_slice(read.filled());
+                        write_packet_to_conn_from_addr(conn, packet, addr);
+                    }
+                }
+            }
+            Err(err) if err.kind() == ErrorKind::TimedOut => {}
+            Err(err) => {
+                log::error!("Could not read udp packet: {}", err);
+                return Poll::Ready(UDPMuxEvent::Error(err));
+            }
+        }
+
+        Poll::Ready(UDPMuxEvent::None)
+    }
+}
+
+fn write_packet_to_conn_from_addr(conn: UDPMuxConn, packet: Vec<u8>, addr: SocketAddr) {
+    tokio::spawn(async move {
+        if let Err(err) = conn.write_packet(&packet, addr).await {
+            log::error!("Failed to write packet: {} (addr={})", err, addr);
+        }
+    });
```

Comment:
`conn.write_packet` writes packet to a buffer, but it's an async function.

> We can remove this tokio::spawn by buffering the packet to be written

you mean storing packets in a map within `UDPMuxNewAddr`?

Reply:
Yeah but looking at the code in more detail, this seems to get pretty convoluted very quickly. One because of all the locks and two because we don't currently remember, whether we got the connection from the address map or the ufrag map but we would have to know that in order to find it again in the next call to `poll` and send the packet.